lib/src/license_detection/license.dart

```diff
@@ -0,0 +1,283 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+part of 'license_detector.dart';
+
+@sealed
+class License {
+  /// SPDX identifier of the license, is empty in case of unknown license.
+  final String identifier;
+
+  /// Original text from the license file.
+  final String content;
+
+  /// Normalized [Token]s created from the original text.
+  final List<Token> tokens;
+
+  /// Map from [Token.value] to their number of occurrences in this license.
+  final Map<String, int> occurrences;
+
+  License._(this.content, this.tokens, this.occurrences, this.identifier);
+
+  factory License.parse(String identifier, String content) {
+    final tokens = tokenize(content);
+    final table = generateFrequencyTable(tokens);
+    return License._(content, tokens, table, identifier);
+  }
+}
+
+String tokensNormalizedValue(Iterable<Token> tokens) {
+  return tokens.map((token) => token.value).join(' ');
+}
+
+/// Representation of n-gram consistent of `n` tokens.
+///
+/// The number of tokens `n` in the n-grams is determined by  confidence threshold for the match.
+/// See: [computeGranularity].
+@sealed
+class NGram {
+  /// Text for which the hash value was generated.
+  final String text;
+
+  /// [CRC-32][1] checksum value generated for text.
+  ///
+  /// [1]: https://en.wikipedia.org/wiki/Cyclic_redundancy_check
+  final int checksum;
+
+  /// Index of the first token in the checksum.
+  final int start;
+
+  /// Index of the last token in the checksum.
+  final int end;
+
+  int get granularity => text.split(' ').length;
+
+  NGram(this.text, this.checksum, this.start, this.end);
+}
+
+/// A [License] instance with generated [nGrams].
+@visibleForTesting
+@sealed
+class LicenseWithNGrams extends License {
```

Comment:
I would make this a composition instead of inheriting from licence:
```suggestion
class LicenseWithNGrams {
  final Licence license;
  final List<NGram> ngrams;
  ....
```

See https://en.wikipedia.org/wiki/Composition_over_inheritance

Reply:
Yeah, we could go either way here...